Optimize away repeated structural schema and CEL DeclType construction in CRD validation

jpbetz opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

This optimizes away repeated structural schema and CEL "decls" construction, which was flagged as a potential blocker for beta for CRD Validation using CEL (https://github.com/kubernetes/enhancements/pull/3266).

#### Which issue(s) this PR fixes:

Fixes: https://github.com/kubernetes/kubernetes/issues/106438

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

Please use the following format for linking documentation:
- KEP: https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/2876-crd-validation-expression-language/README.md

jpbetz committed:
Reuse structural schema and cel decls during CRD validation

jpbetz commented:
@liggitt would you be willing to review? This is one of the optimizations we had listed to do before taking CEL CRD validation to beta (@DangerOnTheRanger will be working on the other one-- eliminating repeated calls to calculate the min size of a schema).

jpbetz commented:
/sig api-machinery
/priority important-soon
/triage accepted

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:
```diff
@@ -780,6 +781,23 @@ type costInfo struct {
 	// definition. A single totalCost is allocated for each validation call and passed through the stack as the
 	// custom resource definition's OpenAPIv3 schema is recursively validated.
 	TotalCost *totalCost
+	// Schema is a lazily populated structural schema for this costInfo node.
+	Schema *structuralschema.Structural
+	// DeclType is a lazily populated CEL declaration representation of Schema of this costInfo node.
+	DeclType *model.DeclType
```
it's a touch weird to put this inside a costInfo type... is there a clearer name for the type if it's going to expand to be used this way?

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:
```diff
@@ -780,6 +781,23 @@ type costInfo struct {
+// LoadSchemaIfAbsent populates the Schema and DeclType fields of costInfo if they are not already set.
+// This makes it possible for costInfos to lazily construct a structural schema and CEL DeclType only for CRD schemas that
+// have validation rules.
+func (c *costInfo) LoadSchemaIfAbsent(schema *apiextensions.JSONSchemaProps, isRoot bool) {
```
is it valid (or if not valid, possible?) for LoadSchemaIfAbsent to be called with different values for schema and isRoot? it's a little weird to ignore params in favor of internal state on subsequent calls

jpbetz replied:
I'm going to try to eliminate this function entirely and simplify the struct lifecycle.

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/validation_test.go:
```diff
@@ -424,6 +425,7 @@ func TestValidateCustomResource(t *testing.T) {
 		},
 		{name: "immutability transition rule",
 			schema: apiextensions.JSONSchemaProps{
+				Type: "object",
```
curious what changed that made these be required?

jpbetz replied:
These tests failed due to https://github.com/kubernetes/kubernetes/blob/d6f4a5d04a93500477dbb1f5df7e9acc42836383/staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go#L106 not matching.  I can't remember why exactly it wasn't observed before, but it has something to do with how we build DeclTypes earlier now and traverse through them instead of building them (possibly many times) when needed.

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:
```diff
@@ -815,14 +833,14 @@ type ruleCost struct {
 // is unbounded, the resulting costInfo's MaxCardinality is also unbounded.
-func (c *costInfo) MultiplyByElementCost(schema *apiextensions.JSONSchemaProps) costInfo {
+func (c *costInfo) CostForChildSchema(schema *apiextensions.JSONSchemaProps) costInfo {
 	result := costInfo{TotalCost: c.TotalCost, MaxCardinality: unbounded}
```
the returned object here does not inherit any lazily computed Schema or DeclType from `costInfo`. I'm pretty disoriented w.r.t. the flow where this is called, but will that make us have to recompute things we expected to only need to be computed once?

jpbetz replied:
I've restructured this so that the flow is much simpler.  Each call these functions does everything needed and the `LoadSchemaIfAbsent` call has been removed.

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:
```diff
@@ -903,6 +968,12 @@ func ValidateCustomResourceDefinitionOpenAPISchema(schema *apiextensions.JSONSch
+	if len(schema.XValidations) > 0 {
+		// This schema will need Schema information to validate the validation rules, so we construct now
+		// so that it can also be used by the validation rules of any child schemas.
+		nodeCostInfo.LoadSchemaIfAbsent(schema, isRoot)
```
are we relying on calls in the right order to populate this data to let us avoid doing work twice? do we have a way to verify how many times we're doing the ~expensive schema calculations so we can make sure any ordering changes in calls don't reintroduce expensive paths?

jpbetz replied:
I've eliminated this call.  The flow should be easier to reason about now.

jpbetz commented on staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go:
```diff
@@ -207,6 +207,7 @@ func WithTypeAndObjectMeta(s *schema.Structural) *schema.Structural {
 		return s
 	}
 	result := &schema.Structural{
+		Items:           s.Items,
```
Found this in a [`TestZeroValueValidation` integration test failure](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/110135/pull-kubernetes-integration/1548133907776409600).  I'm unclear on why this wasn't observed before, but it's a clear error.

jpbetz replied:
Or maybe I'm thinking about this wrong and only object should be allowed at the root?

jpbetz commented:
/retest

jpbetz commented:
@liggitt this is ready for another pass.

jpbetz commented:
/retest

jpbetz commented:
/retest

jpbetz commented:
/retest

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/defaulting/validation.go:
```diff
@@ -70,6 +70,8 @@ func validate(ctx context.Context, pth *field.Path, s *structuralschema.Structur
 		rootSchema = s
 	}
 
+	isRoot := s == rootSchema
```
what does isRoot control? is setting this to `true` correct if this is validating an `XEmbeddedResource` nested down in the depths of a CR (because `rootSchema = s` on line 70 will make this evaluate to true)

liggitt replied:
put another way, does hard-coding this to `true` or to `false` fail any tests? if not, it probably should

jpbetz replied:
If true the CEL "decl" (CELs type information) is guaranteed by `celmodel.WithTypeAndObjectMeta()` to contain kind, apiVersion, metadata.name and metadata.generateName properties.  I believe we want this to be true both for the root object of a CRD and for all `XEmbeddedResource` (the godoc for SchemaDeclType is also written to indicate that we should set it in both cases).

Hard coding either true or false does fail tests (I ended up fixing this because I had introduced a code defect where it became true everywhere and then started tracing through logic.  Setting it false everywhere should also fail some tests that access the metadata fields, I'll verify this).

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:
```diff
@@ -65,27 +65,31 @@ type Validator struct {
 	var propertiesValidators map[string]Validator
 	if s.Items != nil {
-		itemsValidator = validator(s.Items, s.Items.XEmbeddedResource, perCallLimit)
+		itemsValidator = validator(s.Items, s.Items.XEmbeddedResource, declType.ElemType, perCallLimit)
```
we do a nil check on `declType` below but not here... could this panic? if so, is there a test case missing?

jpbetz replied:
I'll drop the below nil check.  See comment on below thread.

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:
```diff
@@ -65,27 +65,31 @@ type Validator struct {
 
-func validator(s *schema.Structural, isResourceRoot bool, perCallLimit uint64) *Validator {
-	compiledRules, err := Compile(s, isResourceRoot, perCallLimit)
+func validator(s *schema.Structural, isResourceRoot bool, declType *model.DeclType, perCallLimit uint64) *Validator {
```
document `declType` must correspond to the schema `s`?

jpbetz replied:
Sounds good. I've added godoc to this function.

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:
```diff
@@ -65,27 +65,31 @@ type Validator struct {
 		for k, prop := range s.Properties {
-			if p := validator(&prop, prop.XEmbeddedResource, perCallLimit); p != nil {
-				propertiesValidators[k] = *p
+			if declType != nil {
+				if f, ok := declType.Fields[k]; ok {
```
I'm curious under what circumstances declType would be nil or would not have a field corresponding to the property `k`

jpbetz replied:
Should only happen if there is a coding error where a schema is converted to incorrect declType, or an incorrect declType is passed to this function.  I'm going to drop the nil checks and favor a panic if this happens.

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/compilation.go:
```diff
@@ -99,10 +99,9 @@ func getBaseEnv() (*cel.Env, error) {
 //  - nil Program, nil Error: The provided rule was empty so compilation was not attempted
 // perCallLimit was added for testing purpose only. Callers should always use const PerCallLimit as input.
-func Compile(s *schema.Structural, isResourceRoot bool, perCallLimit uint64) ([]CompilationResult, error) {
+func Compile(s *schema.Structural, isResourceRoot bool, declType *celmodel.DeclType, perCallLimit uint64) ([]CompilationResult, error) {
```
`isResourceRoot` param looks unused now and could disagree with the construction of the `declType` param... drop the unused param?

also, document `declType` must correspond to `s`?

jpbetz replied:
Yes, dropping isResourceRoot here.  I'll add the godoc in line with the other godoc being added about declType.

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/cel_validation.go:
```diff
@@ -0,0 +1,250 @@
+	return structural, model.SchemaDeclType(structural, isRoot), nil
+}
+
+// RootCelContext constructs CelSchemaContext for the given root schema.
```
nit: consider using `...CEL...` instead of `...Cel...` everywhere?

jpbetz replied:
SGTM. Done.

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/cel_validation.go:
```diff
@@ -0,0 +1,250 @@
+func (c *CelSchemaContext) ChildFieldContext(schema, propSchema *apiextensions.JSONSchemaProps, fieldName string) CelSchemaContext {
+	r := c.childContext(schema)
+	if c.Schema != nil && c.Schema.Properties != nil {
+		if propSchema, ok := c.Schema.Properties[fieldName]; ok {
+			r.Schema = &propSchema
+		}
+		if declField, ok := c.DeclType.Fields[fieldName]; ok {
+			r.DeclType = declField.Type
+		}
```
am I reading correctly that we always use the `schema` param to build childContext, but in this code path, we ignore the `propSchema` parameter? is there a way to have an initialization step and a consumption step, so we don't leave ourselves code paths where we ignore inputs based on cached data?

liggitt replied:
similar comment about sometimes ignoring params in ChildValuesContext and ChildItemsContext functions below

jpbetz replied:
I've split initialization and consumption. This ended up being a substantial change. I think it's a net win since initialization becomes really simple and I was able to a lot of logic to a couple routines which I find easier to reason about. Let me know what you think.

liggitt replied:
I really like the new structure

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/cel_validation.go:
```diff
@@ -0,0 +1,250 @@
+}
+
+// ChildValuesContext constructs and returns a CelSchemaContext for the properties of an object if this CelSchemaContext is an object.
+func (c *CelSchemaContext) ChildValuesContext(schema *apiextensions.JSONSchemaProps) CelSchemaContext {
```
nit: `ChildAdditionalPropertiesContext(...)` might be easier to follow

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/cel_validation.go:
```diff
@@ -0,0 +1,250 @@
+func (c *CelSchemaContext) ChildValuesContext(schema *apiextensions.JSONSchemaProps) CelSchemaContext {
+	r := c.childContext(schema)
+	if c.Schema != nil && c.Schema.AdditionalProperties != nil {
+		if itemsSchema := c.Schema.AdditionalProperties; itemsSchema != nil {
```
might be easier to follow as:

```suggestion
		if additionalPropertiesSchema := c.Schema.AdditionalProperties; additionalPropertiesSchema != nil {
```

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/cel_validation.go:
```diff
@@ -0,0 +1,250 @@
+		}
+		r.DeclType = c.DeclType.ElemType
+	} else {
+		r.initSchemaFields(schema.AdditionalProperties.Schema, false)
```
can schema or schema.AdditionalProperties be nil?

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/cel_validation.go:
```diff
@@ -0,0 +1,250 @@
+	return r
+}
+
+// ChildSchemaContext returns a CelSchemaContext for all child nodes that are not handled by ChildFieldContext,
+// ChildValuesContext or ChildItemsContext.
```
if it's not a property, item, or additionalProperty, what other child nodes are there?

jpbetz replied:
not, allOf, oneOf and anyOf.  I've added to the godoc.

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/cel_validation.go:
```diff
@@ -0,0 +1,250 @@
+// If the conversion fails Schema and DeclType are left nil since x-kubernetes-validations rules
+// are only allowed for valid structural schemas.
+func (c *CelSchemaContext) initSchemaFields(schema *apiextensions.JSONSchemaProps, isRoot bool) {
+	if schema != nil && len(schema.XValidations) > 0 {
+		// As a performance optimization, only perform the conversions to Schema and DeclType types if the schema has
+		// x-kubernetes-validations rules.
+		structural, declType, err := c.Converter(schema, isRoot)
```
does this mean if I have x-kubernetes-validations at the root level and at a level nested 3 levels deep, I'll end up with two declTypes for those two levels?

jpbetz replied:
If the root level has a x-kubernetes-validations, the schema will be converted to a structural+declType and those types will be traversed while looking for deeper x-kubernetes-validations.  I've added a test that ensures that the number of conversions for cases like this is always 1.

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:
```diff
@@ -943,33 +818,33 @@ func ValidateCustomResourceDefinitionOpenAPISchema(schema *apiextensions.JSONSch
 	if len(schema.AllOf) != 0 {
 		for i, jsonSchema := range schema.AllOf {
-			allErrs = append(allErrs, ValidateCustomResourceDefinitionOpenAPISchema(&jsonSchema, fldPath.Child("allOf").Index(i), ssv, false, opts, nodeCostInfo.MultiplyByElementCost(schema))...)
+			allErrs = append(allErrs, ValidateCustomResourceDefinitionOpenAPISchema(&jsonSchema, fldPath.Child("allOf").Index(i), ssv, false, opts, celContext.ChildSchemaContext(schema, &jsonSchema))...)
```
is passing a pointer of the iterator var going to hand the same pointer to all calls? (same question applies to all the other places we're taking pointers to iterator vars)

jpbetz replied:
Turns out that the lifecycle of the pointer is contained in a single iteration of the loop today, but it's quite sketchy to assume that will always be the case.  I'll introduce new vars for all these cases.

liggitt reviewed (commented):
I didn't review the tests yet but wanted to do a sweep in the impl

jpbetz commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/cel_validation.go:
```diff
@@ -0,0 +1,250 @@
+	// has no x-kubernetes-validations rules or if the schema is not a valid structural schema.
+	DeclType *model.DeclType
+
+	// Converter converts a JSON schema into a structuralschema.Structural and model.DeclType.
+	Converter func(schema *apiextensions.JSONSchemaProps, isRoot bool) (*structuralschema.Structural, *model.DeclType, error)
```
Oh yes, I did mean to use the `Converter` type. Fixed.

I've clarified the `Converter` purpose in the godoc.  I added it solely so I could inject a counter when testing to verify how many conversions occur.

jpbetz commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/cel_validation.go:
```diff
@@ -0,0 +1,250 @@
+
+// CelSchemaContext keeps track of data used by x-kubernetes-validations rules for a specific schema node.
+type CelSchemaContext struct {
+	// Schema is a structural schema for this CelSchemaContext node. It is null if the current schema
+	// has no x-kubernetes-validations rules or if the schema is not a valid structural schema.
```
At this level or above.  I've clarified in the godoc.

jpbetz commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/cel_validation.go:
```diff
@@ -0,0 +1,250 @@
+type Converter func(schema *apiextensions.JSONSchemaProps, isRoot bool) (*structuralschema.Structural, *model.DeclType, error)
+
+// DefaultConverter converts using standard conversion utilities.
+var DefaultConverter = func(schema *apiextensions.JSONSchemaProps, isRoot bool) (*structuralschema.Structural, *model.DeclType, error) {
```
Good catch. No, this should be no ever be overridden.  Fixed.

jpbetz commented:
/retest

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/cel_validation.go:
```diff
@@ -0,0 +1,322 @@
+}
+
+// RootCelContext constructs CELSchemaContext for the given root schema.
+func RootCelContext(schema *apiextensions.JSONSchemaProps, converter Converter) CELSchemaContext {
```
if the `converter` param is only for wrapping `DefaultConverter` for unit testing, and the only testing that is using it is in-package, I'd suggest:
1. document the field as being for unit test purposes and set to DefaultConverter otherwise
2. omitting the param from this constructor
3. initializing the field to `DefaultConverter` in this function
4. make `DefaultConverter` private as well, since the only callers are in-package
5. privately stomping the `converter` field after calling `RootCelContext` in validation/validation_test.go

that keeps callers from having to worry about this bit and reduces the surface area

jpbetz replied:
Nice, yeah, that cleans up the surface area.

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/cel_validation.go:
```diff
@@ -0,0 +1,322 @@
+}
+
+// childContext constructs a new CELSchemaContext for the given child schema of the current schema context.
+// extractor optionally provides a way to accessTypeInfo CELTypeInfo of the child from the current schema context's CELTypeInfo.
```
```suggestion
// extractor optionally provides a way to access CELTypeInfo of the child from the current schema context's CELTypeInfo.
```

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/cel_validation.go:
```diff
@@ -0,0 +1,322 @@
+}
+
+// ChildSchemaContext returns a CELSchemaContext for all child nodes that are not handled by ChildPropertiesContext,
+// ChildAdditionalPropertiesContext or ChildItemsContext such as not, allOf, oneOf and anyOf.
```
I'm a little confused why we need this bit... I thought x-kubernetes-validations weren't allowed within not / allOf / oneOf / anyOf ... trying it, I got this:

```
* spec.validation.openAPIV3Schema.properties[a].allOf[0].x-kubernetes-validations: Forbidden: must be empty to be structural
* spec.validation.openAPIV3Schema.properties[a].anyOf[0].x-kubernetes-validations: Forbidden: must be empty to be structural
* spec.validation.openAPIV3Schema.properties[a].not.x-kubernetes-validations: Forbidden: must be empty to be structural
* spec.validation.openAPIV3Schema.properties[a].oneOf[0].x-kubernetes-validations: Forbidden: must be empty to be structural
```

... so there's no rules we'll need to attach cost computations to under these... can we just pass nil for celContext to ValidateCustomResourceDefinitionOpenAPISchema for these?

jpbetz replied:
We don't allow it.  I'll clean this up.

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/cel_validation.go:
```diff
@@ -0,0 +1,322 @@
+		return c.typeInfo
+	}
+
+	// If able to get the type info from the parent's type info, prefer this approach
+	// since it is more efficient.
+	if c.parent != nil {
+		parentTypeInfo := c.parent.TypeInfo()
+		if parentTypeInfo != nil && c.typeInfoAccessor != nil {
+			c.typeInfo = c.typeInfoAccessor.accessTypeInfo(parentTypeInfo)
+			return c.typeInfo
+		}
+	}
```
this was a neat trick, I like this

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/cel_validation.go:
```diff
@@ -0,0 +1,322 @@
+
+	// typeInfo is lazily loaded for schema nodes withinValidationRuleScope and may be
+	// populated one of two possible ways:
+	//   1. Using a typeInfoAccessor to accessTypeInfo it from the parent's type info. This is a cheap operation and should be
```
```suggestion
	//   1. Using a typeInfoAccessor to access it from the parent's type info. This is a cheap operation and should be
```

jpbetz replied:
Grrr, my IDE has some silly notions around how refactors should be applied. Fixed. (I'll dig into IDE settings as well).

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/cel_validation.go:
```diff
@@ -0,0 +1,322 @@
+	// typeInfo is lazily loaded for schema nodes withinValidationRuleScope and may be
+	// populated one of two possible ways:
+	//   1. Using a typeInfoAccessor to accessTypeInfo it from the parent's type info. This is a cheap operation and should be
+	//      used a schema at a higher level already has type info.
```
```suggestion
	//      used when a schema at a higher level already has type info.
```

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:
```diff
@@ -943,33 +818,38 @@ func ValidateCustomResourceDefinitionOpenAPISchema(schema *apiextensions.JSONSch
 
 	if len(schema.AllOf) != 0 {
 		for i, jsonSchema := range schema.AllOf {
-			allErrs = append(allErrs, ValidateCustomResourceDefinitionOpenAPISchema(&jsonSchema, fldPath.Child("allOf").Index(i), ssv, false, opts, nodeCostInfo.MultiplyByElementCost(schema))...)
+			allOfSchema := jsonSchema
+			allErrs = append(allErrs, ValidateCustomResourceDefinitionOpenAPISchema(&allOfSchema, fldPath.Child("allOf").Index(i), ssv, false, opts, celContext.ChildSchemaContext(&allOfSchema))...)
 		}
 	}
 
 	if len(schema.OneOf) != 0 {
 		for i, jsonSchema := range schema.OneOf {
-			allErrs = append(allErrs, ValidateCustomResourceDefinitionOpenAPISchema(&jsonSchema, fldPath.Child("oneOf").Index(i), ssv, false, opts, nodeCostInfo.MultiplyByElementCost(schema))...)
+			oneOfSchema := jsonSchema
+			allErrs = append(allErrs, ValidateCustomResourceDefinitionOpenAPISchema(&oneOfSchema, fldPath.Child("oneOf").Index(i), ssv, false, opts, celContext.ChildSchemaContext(&oneOfSchema))...)
 		}
 	}
 
 	if len(schema.AnyOf) != 0 {
 		for i, jsonSchema := range schema.AnyOf {
-			allErrs = append(allErrs, ValidateCustomResourceDefinitionOpenAPISchema(&jsonSchema, fldPath.Child("anyOf").Index(i), ssv, false, opts, nodeCostInfo.MultiplyByElementCost(schema))...)
+			anyOfSchema := jsonSchema
+			allErrs = append(allErrs, ValidateCustomResourceDefinitionOpenAPISchema(&anyOfSchema, fldPath.Child("anyOf").Index(i), ssv, false, opts, celContext.ChildSchemaContext(&anyOfSchema))...)
 		}
 	}
 
 	if len(schema.Definitions) != 0 {
 		for definition, jsonSchema := range schema.Definitions {
-			allErrs = append(allErrs, ValidateCustomResourceDefinitionOpenAPISchema(&jsonSchema, fldPath.Child("definitions").Key(definition), ssv, false, opts, nodeCostInfo.MultiplyByElementCost(schema))...)
+			definitionSchema := jsonSchema
+			allErrs = append(allErrs, ValidateCustomResourceDefinitionOpenAPISchema(&definitionSchema, fldPath.Child("definitions").Key(definition), ssv, false, opts, celContext.ChildSchemaContext(&definitionSchema))...)
 		}
 	}
```
see question on ChildSchemaContext whether we need to be propagating cel info down into Not/AnyOf/AllOf/OneOf/Definitions recursion... I don't think we do

jpbetz replied:
You're right. It's cleaned up now.

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:
```diff
@@ -983,17 +863,18 @@ func ValidateCustomResourceDefinitionOpenAPISchema(schema *apiextensions.JSONSch
 	if schema.Dependencies != nil {
 		for dependency, jsonSchemaPropsOrStringArray := range schema.Dependencies {
-			allErrs = append(allErrs, ValidateCustomResourceDefinitionOpenAPISchema(jsonSchemaPropsOrStringArray.Schema, fldPath.Child("dependencies").Key(dependency), ssv, false, opts, nodeCostInfo.MultiplyByElementCost(schema))...)
+			allErrs = append(allErrs, ValidateCustomResourceDefinitionOpenAPISchema(jsonSchemaPropsOrStringArray.Schema, fldPath.Child("dependencies").Key(dependency), ssv, false, opts, celContext.ChildSchemaContext(jsonSchemaPropsOrStringArray.Schema))...)
```
dependencies isn't supported, can probably skip propagating cel here

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:
```diff
@@ -1095,20 +976,19 @@ func ValidateCustomResourceDefinitionOpenAPISchema(schema *apiextensions.JSONSch
-		structural, err := structuralschema.NewStructural(schema)
-		if err == nil {
-			compResults, err := cel.Compile(structural, isRoot, cel.PerCallLimit)
+		if celContext.HasValidationRules() && celContext.TypeInfo() != nil {
```
this is inside an `if len(schema.XValidations) > 0 {` block... why do we need `celContext.HasValidationRules()`?

liggitt replied:
if `len(schema.XValidations) > 0`, and `celContext.TypeInfo() == nil`, shouldn't that be an error, rather than silently not compiling/honoring their validation rules? (this looks like a pre-existing issue since we ignored validations when `structuralschema.NewStructural(schema)` returned an error)

jpbetz replied:
Right.  I also found this when following through on some of the above comments about error handling.

jpbetz commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:
```diff
@@ -65,27 +65,32 @@ type Validator struct {
 		for k, prop := range s.Properties {
-			if p := validator(&prop, prop.XEmbeddedResource, perCallLimit); p != nil {
-				propertiesValidators[k] = *p
+			if f, ok := declType.Fields[k]; ok {
```
1. Oh boy. There are a couple problems with the escaping.  I've fixed and added tests to cover the different flows that were broken.
2. If there is a escaped or even an unescapable field, validation should still occur as normal.  Our only rule is that we don't allow CEL to access unescapable fields.  I've added tests covering these cases.  If we otherwise are unable to access a field that exists in the structural schema, that's an internal error.  Since we've moved all the code that could fail in this way to consumption, I believe those should be treated as an internal error (something unexpected has happened internally and we will need to fix whatever went wrong).
3. Desired behavior is:  CEL cannot access the field in rules, but we still traverse it while performing validation. I.e. a rule declared on a schema nested under an unescapable property name still should work as expected. This is fixed and working as expected now (and tested).

jpbetz commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/cel_validation.go:
```diff
@@ -0,0 +1,322 @@
+	// If unable to get the type info from the parent, convert the jsonSchema to type info.
+	// This is expensive for large schemas.
+	typeInfo, err := c.converter(c.jsonSchema, c.parent == nil || c.jsonSchema.XEmbeddedResource)
+	if err == nil {
```
SGTM. Incorporated.

jpbetz commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/cel_validation.go:
```diff
@@ -0,0 +1,322 @@
+}
+
+// ChildPropertyContext constructs and returns a CELSchemaContext for propertyName.
+func (c *CELSchemaContext) ChildPropertyContext(propSchema *apiextensions.JSONSchemaProps, propertyName string) CELSchemaContext {
```
Good catch. This got missed when switching over to a pointer type. Fixing.

jpbetz commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/cel_validation.go:
```diff
@@ -0,0 +1,322 @@
+func (c propertyTypeInfoAccessor) accessTypeInfo(parentTypeInfo *CELTypeInfo) *CELTypeInfo {
+	if parentTypeInfo.Schema.Properties != nil {
+		propSchema := parentTypeInfo.Schema.Properties[c.propertyName]
+		fieldDeclType := parentTypeInfo.DeclType.Fields[c.propertyName]
```
Addressed as part of the clean up pass I did for the above comment.

jpbetz commented:
/retest

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/cel_validation.go:
```diff
@@ -70,37 +72,33 @@ type CELSchemaContext struct {
-	return len(c.jsonSchema.XValidations) > 0
-}
-
 // TypeInfo returns the CELTypeInfo for this CELSchemaContext node.  Returns nil if current level or above does not have
```
nit: suggest "... Returns nil, nil if this CELSchemaContext is nil, or if the current level or above..."

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/cel_validation.go:
```diff
@@ -114,11 +112,11 @@ type CELTypeInfo struct {
-// fails.
+// fails. This should be defaultConverter except in tests where it is useful to wrap it with a converter that tracks
+// how many conversions have been performed.
 type Converter func(schema *apiextensions.JSONSchemaProps, isRoot bool) (*CELTypeInfo, error)
```
nit: can unexport this as well, I think

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/cel_validation.go:
```diff
@@ -130,51 +128,54 @@ func DefaultConverter(schema *apiextensions.JSONSchemaProps, isRoot bool) (*CELT
 	return r
 }
 
 // ChildPropertyContext constructs and returns a CELSchemaContext for propertyName.
```
nit: doc this returns nil if c is nil (applies to the others as well)

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/cel_validation.go:
```diff
@@ -70,37 +72,33 @@ type CELSchemaContext struct {
+			return nil, parentErr
+		}
 		if parentTypeInfo != nil && c.typeInfoAccessor != nil {
 			c.typeInfo = c.typeInfoAccessor.accessTypeInfo(parentTypeInfo)
```
Some `accessTypeInfo` implementations can return nil (like `propertyTypeInfoAccessor#accessTypeInfo`)... don't we need to check for that and fall through to the expensive `c.typeInfo, c.typeInfoErr = c.converter` path in that case?

Do we have a test case for x-kubernetes-validations rules inside a property with an unescapable name? I'm not seeing how that would work with this code as-is (since it looks like it would try to get the typeInfo from the parent, fail because the property name is not escapable, then not fall through to the path below to get the schema for the level with the x-kubernetes-validations rules)

jpbetz replied:
Thank you!  It's a bug and there is a test gap.

I've fleshed out the tests to have all combinations of:

- {unescaped property name ("f"), escaped property name ("f/2"), unescapable property name ("f@2"), items, additionalProperties}
- {parent has schema, parent does not have schema}
- {CRD validation phase, CR validation phase}

It's quite a few more tests, but it catches this bug and I'll verify that the coverage through the new code should now be more comprehensive.

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:
```diff
@@ -976,30 +976,37 @@ func ValidateCustomResourceDefinitionOpenAPISchema(schema *apiextensions.JSONSch
+			typeInfo, err := celContext.TypeInfo()
 			if err != nil {
-				allErrs = append(allErrs, field.InternalError(fldPath.Child("x-kubernetes-validations"), err))
+				allErrs = append(allErrs, field.Invalid(fldPath.Child("x-kubernetes-validations"), nil, fmt.Sprintf("internal error: failed to construct type information for x-kubernetes-validations rules: %s", err)))
```
if we're calling this "internal error:" should we use `field.InternalError`? a `field.Invalid` with a `nil` value makes for a confusing error message (same question on the next error)

jpbetz replied:
My mistake, using `InternalError` everywhere now.

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:
```diff
@@ -82,10 +82,21 @@ func validator(s *schema.Structural, isResourceRoot bool, declType *model.DeclTy
 	if len(s.Properties) > 0 {
 		propertiesValidators = make(map[string]Validator, len(s.Properties))
 		for k, prop := range s.Properties {
```
should we declare a local var to the loop instead of taking the address of `&prop` in a couple places below?

jpbetz replied:
Thanks for catching.

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:
```diff
@@ -82,10 +82,21 @@ func validator(s *schema.Structural, isResourceRoot bool, declType *model.DeclTy
+			} else {
+				// field may be absent from declType if the property name is unescapable, in which case we should convert
+				// the field value type to a DeclType.
+				fieldType = model.SchemaDeclType(&prop, prop.XEmbeddedResource)
```
is it possible for this to return nil? if so, should we `continue`?

jpbetz replied:
It is possible. When we skip CEL validation if the OpenAPIv3 is invalid it should stop happening, but I'd like to have the safety checks regardless. Fixed.

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/cel_validation.go:
```diff
@@ -0,0 +1,326 @@
+	// propertyName is the property name in the parent schema that this schema is declared at.
+	propertyName string
+}
+
```
This is a general comment for these parent property accessors... I don't have a good sense for how paranoid they need to be about checking for nil values in parentTypeInfo or subfields. Can they assume parentTypeInfo is valid and has non-nil schema info / declType or do they need to check every possible nil subfield and return nil so callers can fall back the `converter` path for getting schema info?

jpbetz replied:
The rules are:

- CELTypeInfo must contain non-nil Schema and DeclType fields.
- accessTypeInfo must only be called with a non-nil parent CELTypeInfo

The code is enforcing this. I've updated the godoc to match and to clarify.

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:
```diff
@@ -81,7 +81,8 @@ func validator(s *schema.Structural, isResourceRoot bool, declType *model.DeclTy
 		propertiesValidators = make(map[string]Validator, len(s.Properties))
-		for k, prop := range s.Properties {
+		for k, p := range s.Properties {
+			prop := &p
```
this still isn't quite right... taking the address of the iterator var is the problem (https://go.dev/play/p/wOWJw0tr_LI)

either of these would work:

```go
for k, p := range s.Properties {
  prop := p
  ...
  ... &prop
```

or

```go
for k := range s.Properties {
  prop := s.Properties[k]
  ...
  ... &prop
```

jpbetz replied:
oh yes. Thanks for catching. Fixed.

liggitt commented:
https://github.com/kubernetes/kubernetes/pull/110135#discussion_r928035900 is the only outstanding comment I have. Go ahead and squash once that's done, then lgtm

Thanks for all the work here, the expanded test coverage looks good

jpbetz commented:
> [#110135 (comment)](https://github.com/kubernetes/kubernetes/pull/110135#discussion_r928035900) is the only outstanding comment I have. Go ahead and squash once that's done, then lgtm
> 
> Thanks for all the work here, the expanded test coverage looks good

Fixed, squashed and rebased. Thanks for the review!

liggitt commented:
/lgtm
/approve
